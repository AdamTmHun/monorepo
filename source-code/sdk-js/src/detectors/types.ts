import type { BCP47LanguageTag } from '@inlang/core/languageTag'

type MaybePromise<T> = T | Promise<T>

export type Detector<Parameters extends Array<unknown> = Array<never>> = (
	...parameters: Parameters
) => MaybePromise<BCP47LanguageTag[]>

export type DetectorInitializer<Parameters extends Array<unknown> = Array<never>> = (
	...parameters: Parameters
) => Detector<Parameters>
