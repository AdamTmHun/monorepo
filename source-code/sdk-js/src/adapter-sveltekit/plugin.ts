import { createUnplugin } from "unplugin"

const unplugin = createUnplugin(() => {
	return {
		name: "inlang-sveltekit-adapter",
		buildStart() {
			console.log("Here comes some inlang magic")
		},
	}
})

export const vitePlugin = unplugin.vite
export const rollupPlugin = unplugin.rollup
