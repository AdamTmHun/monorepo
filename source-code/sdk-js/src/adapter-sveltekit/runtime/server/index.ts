export { type SvelteKitServerRuntime, initSvelteKitServerRuntime } from "./runtime.js"
export {
	initHandleWrapper,
	initRootLayoutServerLoadWrapper,
	initServerLoadWrapper,
} from "./wrappers.js"
export {
	referenceLanguage,
	languages,
	getResource,
	reloadResources,
	addRuntimeToLocals,
	getRuntimeFromLocals,
} from "./state.js"
