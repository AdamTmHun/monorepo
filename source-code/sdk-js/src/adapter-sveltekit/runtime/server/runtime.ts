import type { BCP47LanguageTag } from '@inlang/core/languageTag'
import { initRuntimeWithLanguageInformation } from "../../../runtime/index.js"
import { getResource } from "./index.js"

type InitSvelteKitServerRuntimeArgs = {
	sourceLanguageTag: BCP47LanguageTag
	languageTags: BCP47LanguageTag[]
	languageTag: BCP47LanguageTag | undefined
}

export const initSvelteKitServerRuntime = ({
	languageTag,
	sourceLanguageTag,
	languageTags,
}: InitSvelteKitServerRuntimeArgs) => {
	const runtime = initRuntimeWithLanguageInformation({
		readResource: (languageTag: BCP47LanguageTag) => getResource(languageTag),
		sourceLanguageTag,
		languageTags,
	})

	if (languageTag) {
		runtime.loadResource(languageTag)
		runtime.changeLanguageTag(languageTag)
	}

	return runtime
}

// TODO: server should also expose a `route` function

export type SvelteKitServerRuntime = Awaited<ReturnType<typeof initSvelteKitServerRuntime>>
