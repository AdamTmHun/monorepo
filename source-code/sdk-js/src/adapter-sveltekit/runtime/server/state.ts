import type { InlangConfig, InlangConfigModule } from "@inlang/core/config"
import type { RequestEvent } from "@sveltejs/kit"
import { initConfig } from "../../../config/config.js"
import { inlangSymbol } from "../shared/utils.js"
import type { SvelteKitServerRuntime } from "./runtime.js"
import type { LanguageTag } from "@inlang/app"
import type { Message } from '@inlang/messages'

let config: InlangConfig | undefined

export const initState = async (module: InlangConfigModule) => {
	if (!config && !import.meta.env.DEV) {
		try {
			const { languageTags, sourceLanguageTag, resources } = await import("virtual:inlang-static")

			config = {
				sourceLanguageTag,
				languageTags,
				readResources: async () => resources,
				writeResources: async () => undefined,
			} as any as InlangConfig
		} catch {
			/* empty */
		}
	}

	if (!config) {
		config = await initConfig(module)
	}

	await reloadMessages()

	return {
		sourceLanguageTag: config.referenceLanguage,
		languageTags: config.languages,
	}
}

// ------------------------------------------------------------------------------------------------

let _messages: Message[] = []

// TODO: fix resources if needed (add missing Keys, etc.)
export const reloadMessages = async () =>
	(_messages = (await config?.readResources({ config })) as any || [])

export const loadMessages = (languageTag: LanguageTag) =>
	_messages // TODO: filter

// ------------------------------------------------------------------------------------------------

export const addRuntimeToLocals = (
	locals: RequestEvent["locals"],
	runtime: SvelteKitServerRuntime,
) => ((locals as any)[inlangSymbol] = runtime)

export const getRuntimeFromLocals = (locals: RequestEvent["locals"]): SvelteKitServerRuntime =>
	(locals as any)[inlangSymbol]
