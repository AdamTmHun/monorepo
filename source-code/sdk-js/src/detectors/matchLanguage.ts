import type { BCP47LanguageTag } from '@inlang/core/languageTag'
import type { Detector } from "./types.js"

/**
 * A function that takes one or more detected languageTags and matches them with the specified available languageTags.
 * Depending on the passed allowRelated parameter also related languageTags ("en" and "en-GB") are considered a match.
 * Related matches are only considered if no perfect matches can be found.
 * @param detectedLanguages The detected languageTags available for matching. Either an array of strings, a single string or undefined
 * @param languageTags A set of available languageTags (strings) for matching. Insertion order is irrelevant
 * @param allowRelated A boolean specifying whether related languageTags ("en" and "en-US") can be considered if no perfect match can be found in availableLanguages and detected
 * @returns Return string in case of a successful match or otherwise undefined
 */
export const matchLanguage = (
	detectedLanguages: ReturnType<Detector>,
	languageTags: BCP47LanguageTag[],
	allowRelated = true,
) => {
	for (const detectedLanguage of detectedLanguages) {
		// check for perfect match
		for (const languageTag of languageTags) {
			if (languageTag === detectedLanguage) return languageTag
		}
	}
	for (const detectedLanguage of detectedLanguages) {
		if (allowRelated) {
			const relatedLanguages: string[] = languageTags
				.map(
					(languageTag) =>
						(languageTag.startsWith(detectedLanguage + "-") ||
							detectedLanguage.startsWith(languageTag + "-")) &&
						languageTag,
				)
				.filter(Boolean)
				// sort alphabetically
				.sort()
				// unspecific related languageTags should be preferred
				.sort((a, b) => a.split("-").length - b.split("-").length)

			if (relatedLanguages.length) return relatedLanguages[0]
		}
	}

	return undefined
}
