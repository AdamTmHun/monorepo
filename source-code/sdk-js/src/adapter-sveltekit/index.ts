import { standaloneUpdateSdkModuleVersion } from './vite-plugin/config/utils/updateSdkModuleVersion.js'
import { plugin } from "./vite-plugin/plugin.js"

export default plugin
export { standaloneUpdateSdkModuleVersion as updateSdkModuleVersion }
