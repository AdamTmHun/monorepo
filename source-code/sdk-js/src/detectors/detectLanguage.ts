import type { BCP47LanguageTag } from '@inlang/core/languageTag'
import { matchLanguage } from "./matchLanguage.js"
import type { Detector } from "./types.js"

type DetectLanguageParams = {
	sourceLanguageTag: BCP47LanguageTag
	languageTags: BCP47LanguageTag[]
	allowRelated?: boolean
}

/**
 * Takes a set of detection strategies and returns a languageTag found in languageTags.
 * If no matching languageTag can be found, the specified fallback languageTag will be returned.
 * @param params Function parameters
 * @param params.strategies A Set of detectors ordered by precedence (preferred first) (as JS Sets preserve insertion order).
 * These detectors have to be functions that return either an array of strings, a single string or undefined
 * @param params.sourceLanguageTag The languageTag to fallback to if there is no detection match
 * @param params.languageTags Set of languageTags that are available for matching
 * @param params.allowRelated True if also related languageTags shall be considered a match.
 * detectLanguage will then return "en" from languageTags for a detected "en-GB" if no exact match can be found. It can also return "en-GB" from languageTags for a detected "en")
 * @returns The detected languageTag
 */
export const detectLanguage = async (
	{ sourceLanguageTag, languageTags, allowRelated = true }: DetectLanguageParams,
	...detectors: Detector[]
): Promise<BCP47LanguageTag> => {
	const allDetectedLanguages: BCP47LanguageTag[] = []
	for (const detector of detectors) {
		const detectedLanguages = await detector()
		const matchedLanguage = matchLanguage(detectedLanguages, languageTags, false)
		if (matchedLanguage) return matchedLanguage

		allDetectedLanguages.push(...detectedLanguages)
	}

	return (allowRelated && matchLanguage(allDetectedLanguages, languageTags)) || sourceLanguageTag
}
