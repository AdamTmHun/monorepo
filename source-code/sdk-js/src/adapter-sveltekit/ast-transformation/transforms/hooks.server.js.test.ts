import { describe, expect, test } from "vitest"
import { transformHooksServerJs } from "./hooks.server.js.js"

describe("hooks.server.js.ts", () => {
	const baseConfig = {
		isStatic: true,
		srcFolder: "",
		rootRoutesFolder: "",
		hasAlreadyBeenInitialized: true,
	}
	const requiredImports = `import { replaceLanguageInUrl } from "@inlang/sdk-js/adapter-sveltekit/shared";
import { redirect } from "@sveltejs/kit";
import { initAcceptLanguageHeaderDetector } from "@inlang/sdk-js/detectors/server";
import { initHandleWrapper } from "@inlang/sdk-js/adapter-sveltekit/server";`
	test("Insert into empty file with no options", () => {
		const code = ""
		const config = {
			languageInUrl: true,
			...baseConfig,
		}
		const transformed = transformHooksServerJs(config, code)
		const expected = `${requiredImports}
export const handle = initHandleWrapper({
  getLanguage: ({ url }) => url.pathname.split("/")[1]
}).wrap(async ({ event, resolve }) => resolve(event));`
		expect(transformed).toBe(expected)
	})
})
