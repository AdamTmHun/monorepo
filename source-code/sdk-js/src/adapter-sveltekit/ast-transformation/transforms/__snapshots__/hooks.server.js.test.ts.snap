// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformHooksServerJs > isStatic 1`] = `
"import { replaceLanguageInUrl } from \\"@inlang/sdk-js/adapter-sveltekit/shared\\";
import { redirect } from \\"@sveltejs/kit\\";
import { initAcceptLanguageHeaderDetector } from \\"@inlang/sdk-js/detectors/server\\";
import { initHandleWrapper } from \\"@inlang/sdk-js/adapter-sveltekit/server\\";
export const handle = initHandleWrapper({
  getLanguage: () => undefined
}).wrap(async ({ event, resolve }) => resolve(event));"
`;

exports[`transformHooksServerJs > languageInUrl and isStatic 1`] = `
"import { replaceLanguageInUrl } from \\"@inlang/sdk-js/adapter-sveltekit/shared\\";
import { redirect } from \\"@sveltejs/kit\\";
import { initAcceptLanguageHeaderDetector } from \\"@inlang/sdk-js/detectors/server\\";
import { initHandleWrapper } from \\"@inlang/sdk-js/adapter-sveltekit/server\\";
export const handle = initHandleWrapper({
  getLanguage: ({ url }) => url.pathname.split(\\"/\\")[1]
}).wrap(async ({ event, resolve }) => resolve(event));"
`;
